### PR TITLE
remove redundant codes of invoking paramFetch before preprocess_initplans

### PR DESCRIPTION
**Pull request by SmartKeyerror**
ParamListInfo structures are used to pass parameters into the executor
for parameterized plans.  We support two basic approaches to supplying
parameter values, the "static" way and the "dynamic" way.

In the dynamic approach, all access to parameter values is done through
hook functions found in the ParamListInfo struct. The params will be
lazily evaluated in Executor. And more specifically, dynamic approach is
designed for plpgsql, there has a complete mechanism internally to call 
the `plpgsql_param_fetch` method to obtain dynamic parameters from 
the entrance of `plpgsql_call_handler()`.

In conclusion, we don't need to explicitly call the `paramFetch` method 
to get dynamic parameters before `preprocess_initplans`, but get them 
during the actual execution.

No more tests need, already have regression tests will go through the 
deleted code branch, such as `plpgsql_cache`, `plpgsql`, `case_gp`, etc. 
So no need to add more tests for this PR.

### CODE REVIEW

**Comment by Aegeaner** [comment]
After this change, when will the dynamic parameters be actually fetched, during the expression evaluation stage?

**Comment by SmartKeyerror** [comment]
> After this change, when will the dynamic parameters be actually fetched, during the expression evaluation stage?

This will be processed by plpgsql mechanism to call the `plpgsql_param_fetch` method to obtain dynamic parameters from the entrance of `plpgsql_call_handler()`, it's the sub-procedure of `ExecProcNode`.

Actually, the deleted codes by this PR is not really get dynamic params with `paramFetch`, the canonical procedure is:

```
ParamExternData *prm;
ParamListInfo externParams
ParamExternData prmdata;

if (externParams->paramFetch != NULL)
    prm = externParams->paramFetch(externParams, i + 1, false, &prmdata);
else
    prm = &externParams->params[i];
```

We need to receive the function return value or use the pointer `ParamExternData *prm` that we passed to the function.

**Comment by SmartKeyerror** [comment]
> After this change, when will the dynamic parameters be actually fetched, during the expression evaluation stage?

This will be processed by plpgsql mechanism to call the `plpgsql_param_fetch` method to obtain dynamic parameters from the entrance of `plpgsql_call_handler()`, it's the sub-procedure of `ExecProcNode`.

Actually, the deleted codes by this PR is not really get dynamic params with `paramFetch`, the canonical procedure is:

```
ParamExternData *prm;
ParamListInfo externParams
ParamExternData prmdata;

if (externParams->paramFetch != NULL)
    prm = externParams->paramFetch(externParams, i + 1, false, &prmdata);
else
    prm = &externParams->params[i];
```

We need to receive the function return value or use the pointer `ParamExternData *prm` that we passed to the function.

File: src/backend/executor/execMain.c [hunk 1]
```diff
@@ -603,31 +603,6 @@ standard_ExecutorStart(QueryDesc *queryDesc, int eflags)
 			 */
 			if (list_length(queryDesc->plannedstmt->paramExecTypes) > 0)
 			{
```
Comment (haolinw) [inline src/backend/executor/execMain.c:605]: For coding style, one line no braces under if condition.

Comment (SmartKeyerror) [reply src/backend/executor/execMain.c:605]: seems not necessary? The codes below in this file do not obey this rule too.

Comment (SmartKeyerror) [reply src/backend/executor/execMain.c]: Changed.

**Review by haolinw — APPROVED** [review]

**Comment by adam8157** [comment]
Does this conflict with the comment of `plpgsql_param_fetch()`?
```
/*
 * plpgsql_param_fetch		paramFetch callback for dynamic parameter fetch
 *
 * We always use the caller's workspace to construct the returned struct.
 *
 * Note: this is no longer used during query execution.  It is used during
 * planning (with speculative == true) and when the ParamListInfo we supply
 * to the executor is copied into a cursor portal or transferred to a
 * parallel child process.
 */
```

**Comment by SmartKeyerror** [comment]
> Does this conflict with the comment of `plpgsql_param_fetch()`?
> 
> ```
> /*
>  * plpgsql_param_fetch		paramFetch callback for dynamic parameter fetch
>  *
>  * We always use the caller's workspace to construct the returned struct.
>  *
>  * Note: this is no longer used during query execution.  It is used during
>  * planning (with speculative == true) and when the ParamListInfo we supply
>  * to the executor is copied into a cursor portal or transferred to a
>  * parallel child process.
>  */
> ```

It seems that the comments are confusing, "the query execution", as I think, is the planner stage of plpgsql, in PostgreSQL 12.6, we can get below backtrace:

```
#0  plpgsql_param_fetch (params=0xa00000000, paramid=0, speculative=false, prm=0x7ffe16bbc7a3) at pl_exec.c:6374
#1  0x0000555e2cbe1fb4 in eval_const_expressions_mutator (node=0x555e2dfcaa40, context=0x7ffe16bbd9e0) at clauses.c:2419
#2  0x0000555e2cb3d58d in expression_tree_mutator (node=0x555e2dfcb6a8, mutator=0x555e2cbe1eaf <eval_const_expressions_mutator>, context=0x7ffe16bbd9e0) at nodeFuncs.c:3061
#3  0x0000555e2cbe4c0e in simplify_function (funcid=43, result_type=2275, result_typmod=-1, result_collid=0, input_collid=0, args_p=0x7ffe16bbd1c0, funcvariadic=false, process_args=true, allow_non_const=true, 
    context=0x7ffe16bbd9e0) at clauses.c:3983
#4  0x0000555e2cbe2da5 in eval_const_expressions_mutator (node=0x555e2dfcaa08, context=0x7ffe16bbd9e0) at clauses.c:2882
#5  0x0000555e2cb3d58d in expression_tree_mutator (node=0x555e2dfca928, mutator=0x555e2cbe1eaf <eval_const_expressions_mutator>, context=0x7ffe16bbd9e0) at nodeFuncs.c:3061
#6  0x0000555e2cbe4c0e in simplify_function (funcid=67, result_type=16, result_typmod=-1, result_collid=0, input_collid=100, args_p=0x7ffe16bbd980, funcvariadic=false, process_args=true, allow_non_const=true, 
    context=0x7ffe16bbd9e0) at clauses.c:3983
#7  0x0000555e2cbe24c1 in eval_const_expressions_mutator (node=0x555e2dfca8d0, context=0x7ffe16bbd9e0) at clauses.c:2582
#8  0x0000555e2cbe1e21 in eval_const_expressions (root=0x555e2dfcb158, node=0x555e2dfca8d0) at clauses.c:2326
#9  0x0000555e2cbbfc3a in preprocess_expression (root=0x555e2dfcb158, expr=0x555e2dfca8d0, kind=0) at planner.c:1087
#10 0x0000555e2cbbfd69 in preprocess_qual_conditions (root=0x555e2dfcb158, jtnode=0x555e2dfca818) at planner.c:1149
#11 0x0000555e2cbbf4c8 in subquery_planner (glob=0x555e2dfcaba8, parse=0x555e2dfca3e8, parent_root=0x0, hasRecursion=false, tuple_fraction=0) at planner.c:792
#12 0x0000555e2cbbe7d1 in standard_planner (parse=0x555e2dfca3e8, cursorOptions=256, boundParams=0x555e2dfd23e8) at planner.c:406
#13 0x0000555e2cbbe574 in planner (parse=0x555e2dfca3e8, cursorOptions=256, boundParams=0x555e2dfd23e8) at planner.c:275
#14 0x0000555e2ccdce01 in pg_plan_query (querytree=0x555e2dfca3e8, cursorOptions=256, boundParams=0x555e2dfd23e8) at postgres.c:878
#15 0x0000555e2ccdcf36 in pg_plan_queries (querytrees=0x555e2dfca388, cursorOptions=256, boundParams=0x555e2dfd23e8) at postgres.c:968
#16 0x0000555e2ce5abc8 in BuildCachedPlan (plansource=0x555e2dfa11f8, qlist=0x555e2dfca388, boundParams=0x555e2dfd23e8, queryEnv=0x0) at plancache.c:933
#17 0x0000555e2ce5b22c in GetCachedPlan (plansource=0x555e2dfa11f8, boundParams=0x555e2dfd23e8, useResOwner=true, queryEnv=0x0) at plancache.c:1214
#18 0x0000555e2cb18240 in _SPI_execute_plan (plan=0x555e2df9d5d8, paramLI=0x555e2dfd23e8, snapshot=0x0, crosscheck_snapshot=0x0, read_only=false, fire_triggers=true, tcount=1) at spi.c:2216
#19 0x0000555e2cb1515c in SPI_execute_plan_with_paramlist (plan=0x555e2df9d5d8, params=0x555e2dfd23e8, read_only=false, tcount=1) at spi.c:576
#20 0x00007fa16b26ef47 in exec_stmt_execsql (estate=0x7ffe16bbe280, stmt=0x555e2dfa1bc0) at pl_exec.c:4216
#21 0x00007fa16b26a53f in exec_stmt (estate=0x7ffe16bbe280, stmt=0x555e2dfa1bc0) at pl_exec.c:2041
#22 0x00007fa16b26a234 in exec_stmts (estate=0x7ffe16bbe280, stmts=0x555e2dfa1c40) at pl_exec.c:1932
#23 0x00007fa16b26a0c1 in exec_stmt_block (estate=0x7ffe16bbe280, block=0x555e2dfa1d18) at pl_exec.c:1873
#24 0x00007fa16b26a33e in exec_stmt (estate=0x7ffe16bbe280, stmt=0x555e2dfa1d18) at pl_exec.c:1965
#25 0x00007fa16b2678d5 in plpgsql_exec_function (func=0x555e2dee4d18, fcinfo=0x555e2dfd43f8, simple_eval_estate=0x0, atomic=true) at pl_exec.c:597
#26 0x00007fa16b2618f6 in plpgsql_call_handler (fcinfo=0x555e2dfd43f8) at pl_handler.c:267
#27 0x0000555e2caceed6 in ExecMakeTableFunctionResult (setexpr=0x555e2dfc0808, econtext=0x555e2dfc06b8, argContext=0x555e2dfd42e0, expectedDesc=0x555e2dfc1020, randomAccess=false) at execSRF.c:234
#28 0x0000555e2cae5bb3 in FunctionNext (node=0x555e2dfc05a0) at nodeFunctionscan.c:95
#29 0x0000555e2cace53b in ExecScanFetch (node=0x555e2dfc05a0, accessMtd=0x555e2cae5afe <FunctionNext>, recheckMtd=0x555e2cae5efa <FunctionRecheck>) at execScan.c:133
#30 0x0000555e2cace5b4 in ExecScan (node=0x555e2dfc05a0, accessMtd=0x555e2cae5afe <FunctionNext>, recheckMtd=0x555e2cae5efa <FunctionRecheck>) at execScan.c:183
#31 0x0000555e2cae5f50 in ExecFunctionScan (pstate=0x555e2dfc05a0) at nodeFunctionscan.c:270
#32 0x0000555e2cacc04d in ExecProcNodeFirst (node=0x555e2dfc05a0) at execProcnode.c:445
#33 0x0000555e2cac062c in ExecProcNode (node=0x555e2dfc05a0) at ../../../src/include/executor/executor.h:242
#34 0x0000555e2cac3029 in ExecutePlan (estate=0x555e2dfc0348, planstate=0x555e2dfc05a0, use_parallel_mode=false, operation=CMD_SELECT, sendTuples=true, numberTuples=0, direction=ForwardScanDirection, 
    dest=0x555e2dfce0a8, execute_once=true) at execMain.c:1632
#35 0x0000555e2cac0cb5 in standard_ExecutorRun (queryDesc=0x555e2df978d8, direction=ForwardScanDirection, count=0, execute_once=true) at execMain.c:350
#36 0x0000555e2cac0ad2 in ExecutorRun (queryDesc=0x555e2df978d8, direction=ForwardScanDirection, count=0, execute_once=true) at execMain.c:294
#37 0x0000555e2cce4015 in PortalRunSelect (portal=0x555e2df1f4f8, forward=true, count=0, dest=0x555e2dfce0a8) at pquery.c:938
#38 0x0000555e2cce3c33 in PortalRun (portal=0x555e2df1f4f8, count=9223372036854775807, isTopLevel=true, run_once=true, dest=0x555e2dfce0a8, altdest=0x555e2dfce0a8, completionTag=0x7ffe16bbeab0 "")
    at pquery.c:779
#39 0x0000555e2ccdd400 in exec_simple_query (query_string=0x555e2deb8aa8 "select * from cache_test(1);") at postgres.c:1215
#40 0x0000555e2cce1c57 in PostgresMain (argc=1, argv=0x555e2dee3690, dbname=0x555e2dee34d0 "postgres", username=0x555e2deb5368 "smart") at postgres.c:4271
```

We can find the invocation of function `ExecProcNode()` is still under the process of `ExecProcNode()`.

**Review by kainwen — APPROVED** [review]
At least this part of the code does nothing: it just discards the result. So to remove it has no risk.

Besides, I cannot make a case that InitPlan refers to an external param.

Before merging, can we spend a little time to find the commit introducing this to see if there is any clue?
